src/ecmult_const_impl.h

```diff
@@ -228,4 +228,127 @@ static void secp256k1_ecmult_const(secp256k1_gej *r, const secp256k1_ge *a, cons
+     * phi_u^{-1} to get back to secp256k1.
+     *
+     * Switching to Jacobian coordinates, note that phi_u applied to (X, Y, Z) is simply
+     * (X, Y, Z/u). Thus, if we want to compute (X1, Y1, Z) + (X2, Y2, Z), with identical Z
+     * coordinates, we can use phi_Z to transform it to (X1, Y1, 1) + (X2, Y2, 1) on an isomorphic
+     * curve where the affine addition formula can be used instead.
+     * If (X3, Y3, Z3) = (X1, Y1) + (X2, Y2) on that curve, then our answer on secp256k1 is
+     * (X3, Y3, Z3*Z).
+     *
+     * This is the effective affine technique: if we have a linear combination of group elements
+     * to compute, and all those group elements have the same Z coordinate, we can simply pretend
+     * that all those Z coordinates are 1, perform the computation that way, and then multiply the
+     * original Z coordinate back in.
+     *
+     * The technique works on any a=0 short Weierstrass curve. It is possible to generalize it to
+     * other curves too, but there the isomorphic curves will have different 'a' coefficients,
+     * which typically does affect the group laws.
+     *
+     *
+     * === Avoiding the square root for x-only point multiplication ===
+     *
+     * In this function, we want to compute the X coordinate of q*(n/d, y), for
+     * y = +-sqrt((n/d)^3 + 7).
+     *
+     * Let g = y^2*d^3 = n^3 + 7*d^3. This also means y = +-sqrt(g/d^3). Further note that
+     * sqrt(d*g) must exist if the input was valid, as d*g = y^2*d^4 = (y*d^2)^2.
+     *
+     * The input point (n/d, y) also has Jacobian coordinates:
+     *   (n/d, y, 1)
+     * = (n/d * sqrt^2(d*g), y * sqrt^3(d*g), sqrt(d*g))
+     * = (n/d * d*g, y * +-sqrt(d^3*g^3), sqrt(d*g))
+     * = (n/d * d*g, +-sqrt(g/d^3) * +-sqrt(d^3*g^3), sqrt(d*g))
+     * = (n*g, +-sqrt(g/d^3 * d^3*g^3), sqrt(d*g))
+     * = (n*g, +-sqrt(g^4), sqrt(d*g))
+     * = (n*g, +-g^2, sqrt(d*g))
+     *
+     * Both signs for the Y coordinate are valid (they satisfy Y^2 = X^3 + 7Z^6), and both
+     * cases correspond to affine X coordinate n/d. We choose the (n*g, g^2, sqrt(d*g)) version.
+     *
+     * Now switch to the effective affine curve using phi_{sqrt(d*g)}, where the input point has
+     * coordinates (n*g, g^2). Compute (X, Y, Z) = q*(n*g, g^2) there.
+     *
+     * Back on secp256k1, that means q*(n*g, g^2, sqrt(d*g)) = (X, Y, Z*sqrt(d*g)). This last
+     * point has affine X coordinate X/(d*g*Z^2). Determining the affine Y coordinate would involve
+     * a square root, but as long as we only care about the resulting X coordinate, no square root
+     * is needed anywhere in this computation.
+     */
+
+    secp256k1_fe g, i;
+    secp256k1_ge p;
+    secp256k1_gej rj;
+
+    /* Compute g = (n^3 + B*d^3). */
+    secp256k1_fe_sqr(&g, n);
+    secp256k1_fe_mul(&g, &g, n);
+    if (d) {
+        secp256k1_fe b;
+        secp256k1_fe_sqr(&b, d);
+        secp256k1_fe_mul_int(&b, SECP256K1_B);
```

Comment:
`b` now has magnitude `SECP256K1_B == 7`, which happens to be just below `8`, the maximum for the following `fe_mul` call.

So, we better don't switch to a different curve in the future. ;) On a more serious note, I don't know how much our code depends on B, but I think it's a good idea to avoid introducing further dependencies on the B in the code base. 

If you agree, we could do this: 

```suggestion
        VERIFY_CHECK(SECP256K1_B <= 8); /* magnitude of b will be <= 8 after the next call */
        secp256k1_fe_mul_int(&b, SECP256K1_B);
```

Of course, the same is already checked now in `fe_verify` but this makes it a bit more explicit.